### PR TITLE
CA-325582: Report dummy stats for unplugged SRs

### PR DESCRIPTION
**Pull request by psafont**
Must be merged along https://github.com/xapi-project/xcp-rrdd/pull/101

This is required because SR metrics are being reported as host ones: when creating and deleting SRs their metrics data-sources are kept and accumulate over time.
To avoid an accompanying change in xcp-rrdd deletes rrds with no corresponding active data-sources for host metrics.

To avoid the deletion of metrics of unplugged PBDS, we make an extra effort to report them.
This means querying xapi for unplugged PBDs and stuffing metrics with unknown values to not upset the existing, valid metrics in the SRs.

### CODE REVIEW

File: src/rrdp-iostat/rrdp_iostat.ml [hunk 1]
```diff
@@ -142,11 +142,9 @@ module Iostat = struct
...
     (* Now read the values out of dev_values_map for devices for which we have data *)
     Listext.List.filter_map (fun dev ->
-        if not (Hashtbl.mem dev_values_map dev)
-        then None
-        else
-          let values = Hashtbl.find dev_values_map dev in
-          Some (dev, values)
+        match Hashtbl.find_opt dev_values_map dev with
+        | None -> None
+        | Some values -> Some (dev, values)
```
Comment (lippirk) [inline src/rrdp-iostat/rrdp_iostat.ml]: `option` is a functor, so you can map:

`Hashtbl.find_opt dev_values_map dev |> Option.map (fun v -> (dev, v))`

Comment (psafont) [reply src/rrdp-iostat/rrdp_iostat.ml]: > Since 4.08
;(

File: src/rrdp-iostat/rrdp_iostat.ml [hunk 2]
```diff
@@ -293,8 +303,11 @@ let exec_tap_ctl () =
   minor_to_sr_and_vdi
 
 let minor_of_tdX_unsafe tdX =
-  int_of_string (Unixext.file_lines_fold (fun acc l -> acc ^ (List.nth (Xstringext.String.split ':' l) 1))
-                   "" ("/sys/block/" ^ tdX ^ "/dev"))
+  Unixext.read_lines ("/sys/block/" ^ tdX ^ "/dev")
+  |> List.hd
+  |> Astring.String.cuts ~sep:":"
+  |> fun l -> List.nth l 1
```
[the remark this answers is not in the document]
Comment (psafont) [reply src/rrdp-iostat/rrdp_iostat.ml]: This file contains the device number in the form of `major:minor`. It seems a fundamental part of how the kernel identifies block devices ao I don't think we need to protect against failure.

[the remark this answers is not in the document]
Comment (lindig) [reply src/rrdp-iostat/rrdp_iostat.ml]: This could be parsed with `sscanf`

[the remark this answers is not in the document]
Comment (psafont) [reply src/rrdp-iostat/rrdp_iostat.ml]: You're right, I think it also applies to some of the other files as well.

**Comment by psafont** [comment]
This is failing because the daemon gets started before xapi. The daemon is the unable to recover from that failure.

**Comment by psafont** [comment]
Superseded by https://github.com/xapi-project/xen-api/pull/3989 and https://github.com/xapi-project/xcp-rrdd/pull/106

In comparison this approached had the potential to regress performance with the new connection to xapi.

**Pull request closed without merging**